util/dvsim/NcLauncher.py

```diff
@@ -86,49 +85,49 @@ def _do_launch(self):
             # Interactive: Set RUN_INTERACTIVE to 1
             exports['RUN_INTERACTIVE'] = '1'
-            interactive_flags = ['-I']
-            cmd_arr = self.get_submit_cmd(interactive_flags)
-            log.log(VERBOSE, '[Executing]:\n{}\n\n'.format(self.deploy.cmd))
+            cmd_arr = self.get_submit_cmd(['-I'])
+            buf_size = 1
+            std_out = subprocess.PIPE
+            std_err = subprocess.STDOUT
+        else:
+            cmd_arr = self.get_submit_cmd(['-wl'])
+            buf_size = 4096
+            std_out = fd
+            std_err = fd
+
+        log.log(VERBOSE, '[Submitting]:\n{}\n\n'.format(' '.join(cmd_arr)))
+
+        try:
             self.process = subprocess.Popen(cmd_arr,
-                                            stdin=None,
-                                            stdout=None,
-                                            stderr=subprocess.STDOUT,
-                                            # string mode
+                                            bufsize=buf_size,
                                             universal_newlines=True,
+                                            pass_fds = [fd],
+                                            stdout=std_out,
+                                            stderr=std_err,
                                             env=exports,
                                             cwd=self.deploy.odir)
-
-            # Wait until the process exits
-            self.process.wait()
-        else:
-            try:
-                interactive_flags = ['-nolog', '-wl']
-                cmd_arr = self.get_submit_cmd(interactive_flags)
-                # Using file open instead of with open as
-                # it is being using in the subprocess.Popen call
-                # which returns immediately after launching the cmd
-                # and we want the file to remain open throughout the process
-                f = open(self.deploy.get_log_path(),
-                         'w',
-                         encoding='UTF-8',
-                         errors='surrogateescape')
-                f.write('[Executing]:\n{}\n\n'.format(self.deploy.cmd))
-                f.flush()
-                self.process = subprocess.Popen(cmd_arr,
-                                                bufsize=4096,
-                                                universal_newlines=True,
-                                                stdout=f,
-                                                stderr=f,
-                                                env=exports,
-                                                cwd=self.deploy.odir)
-            except subprocess.SubprocessError as e:
-                raise LauncherError(f'IO Error: {e}\n'
-                                    f'See {self.deploy.get_log_path()}')
-            finally:
-                self._close_process()
+            if self.deploy.sim_cfg.interactive:
+                for line in self.process.stdout:
```

Comment:
If I understand correctly:
- In non-interactive mode, the subprocess just writes to `fd`.
- In interactive mode, the subprocess writes to some pipe that comes back to the Python process, which then does a "tee operation" and writes the data to the `fd` and also `stdout`.

I guess this makes sense. If you do the rewording I'm suggesting above, we'll have a file object floating around, so won't need the explicit call to `encode`. Maybe that would be marginally simpler?

Reply:
done